File: java/src/P030/SubstringwithConcatenationofAllWords/SubstringwithConcatenationofAllWords.java
package P030.SubstringwithConcatenationofAllWords;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class SubstringwithConcatenationofAllWords {
    public static List<Integer> findSubstring(String s, String[] words) {
        List<Integer> res = new ArrayList<>();
        if (words == null || words.length == 0 || s.length() == 0) return res;
        int wordLen = words[0].length();
        int numWord = words.length;
        int windowLen = wordLen * numWord;
        int sLen = s.length();
        HashMap<String, Integer> map = new HashMap<>();
        for (String word : words) map.put(word, map.getOrDefault(word, 0) + 1);

        // Run wordLen scans
        for (int i = 0; i < wordLen; i++) {
            HashMap<String, Integer> curMap = new HashMap<>();
            // Move window in step of wordLen
            for (int j = i, count = 0, start = i; j + wordLen <= sLen; j += wordLen) {
                // count: number of exceeded occurences in current window
                // start: start index of current window of size windowLen
                if (start + windowLen > sLen) break;
                String word = s.substring(j, j + wordLen);
                if (!map.containsKey(word)) {
                    curMap.clear();
                    count = 0;
                    start = j + wordLen;
                } else {
                    // Remove previous word of current window
                    if (j == start + windowLen) {
                        String preWord = s.substring(start, start + wordLen);
                        start += wordLen;
                        int val = curMap.get(preWord);
                        if (val == 1) curMap.remove(preWord);
                        else curMap.put(preWord, val-1);
                        if (val - 1 >= map.get(preWord)) count--;
                    }
                    // Add new word
                    curMap.put(word, curMap.getOrDefault(word, 0) + 1);

                    // More than expected, increase count
                    if (curMap.get(word) > map.get(word)) count++;

                    // Check if current window valid
                    if (count == 0 && start + windowLen == j + wordLen) {
                        res.add(start);
                    }
                }

            }
        }
        return res;
    }

    public static void main(String[] args) {
        String s = "barfoothefoobarman";
        String[] words = new String[] {"foo", "bar"};
        System.out.println(SubstringwithConcatenationofAllWords.findSubstring(s, words));
    }
}

File: python/P030_Substring_with_Concatenation_of_All_Words/Substring with Concatenation of All Words.py
# -*- coding: utf-8 -*-
class Solution(object):
    def findSubstring(self, s, words):
        """
        :type s: str
        :type words: List[str]
        :rtype: List[int]
        """
        from collections import Counter
        from collections import defaultdict

        res = []
        if not words or len(words[0]) == 0 or len(s) == 0:
            return res
        word_len = len(words[0])
        num_words = len(words)
        window_len = word_len * num_words
        word_map = Counter(words)

        for i in range(word_len):
            cur_map = defaultdict(int)
            start, count = i, 0
            for j in range(i, len(s)-word_len+1, word_len):
                word = s[j:j+word_len]
                if word in word_map:
                    cur_map[word] += 1
                    count += 1
                    while cur_map[word] > word_map[word]:
                        cur_map[s[start:start+word_len]] -= 1
                        start += word_len
                        count -= 1
                    if count == num_words:
                        res.append(start)
                else:
                    cur_map.clear()
                    count = 0
                    start = j + word_len
        return res